### PR TITLE
EPPT-2991 Add _replace_mask_values method to CubeCombiner to unmask values in a masked dataset when the replace_masked_values_with parameter is given.

### PR DESCRIPTION
**Pull request by robertplatt-mo**
EPPT-2991 Add _replace_mask_values method to CubeCombiner to unmask values in a masked dataset when the replace_masked_values_with parameter is given.

Description

During development of https://metoffice.atlassian.net/browse/EPPT-2991 we identified a need for mask handling where the masked data could be treated as zero to enable calculations to take place without the masked area persisting into downstream datasets.  Adding the _replace_mask_values method to CubeCombiner enables this behaviour when the replace_masked_values_with parameter is given.

Testing:

Duplicated the test_with_mask test to create additional tests for the default behaviour and then test the _replace_mask_values behaviour with the `add` and `multiply` operators for different vlaues of `replace_masked_values_with`.

Upon realising that the existing masked_add operator currently retains the mask wherever the input masks are overlapping, added a test for masked_add to demonstrate this behaviour and for the Test_process tests to demonstrate the opposite when _replace_masked_values_with is used.

- [x] Ran tests and they passed OK
- [x] Added new tests for the new feature(s)

CLA

- [x] If a new developer, signed up to CLA

### CODE REVIEW

File: improver/cube_combiner.py [hunk 1]
```diff
@@ -170,6 +173,11 @@ def masked_add(
     Returns:
         numpy.ma.MaskedArray:
             The sum of the two masked arrays with masked points treated as 0.
+
+    Warning:
+        The masked_add function will retain the mask wherever the input masks
+        are overlapping.
+
```
Comment (robertplatt-mo) [inline improver/cube_combiner.py:180]: I added these warnings because I did some digging in to the masked_add function to see whether it was a distinct method that still requires it's own function or whether it can be handled under the `_zero_mask_values` method when `preserve_mask_values=False`. The methods do differ but I'm not entirely convinced this is intentional.

I suspect that this overlapping issue wasn't a concern in the original work (masking land and sea) because when masks are created to distinguish land and sea they are most likely created from the same source coastline (so overlaps should not happen).  That said I'm struggling to think of a use case for a scenario where we retain only overlapping masks rather than allowing them to be visible as zeros. Should we be raising a task to deprecate this function and use the new method instead or is there a use case I'm unaware of that can be included in the docstrings?

The background to the masked_add function was task [EPPT1305](https://metoffice.atlassian.net/browse/EPPT-1305) and the work was on behalf of the [stability adjusted wind gust wf](https://metoffice.sharepoint.com/:p:/r/sites/ProgrammesEnhancingPostProcessingProject/_layouts/15/Doc.aspx?sourcedoc=%7B92D3264A-6975-4CE9-806A-DA67FA4ED746%7D&file=stability%20adjusted%20wind%20gust%20workflows.pptx&action=edit&mobileredirect=true). According to page 6 of the linked workflow plan that work may ultimately not have been needed.

File: improver/cube_combiner.py [hunk 2]
```diff
@@ -342,6 +364,23 @@ def _add_cell_method(self, cube: Cube) -> None:
             )
             raise ValueError(msg)
 
+    def _zero_mask_values(self, cube_list: Union[List[Cube], CubeList]) -> Cube:
```
Comment (Anzerkhan27) [inline improver/cube_combiner.py]: The return type hint should be `List[Cube]`  since this method returns `new_cube_list`.

File: improver/cube_combiner.py (hunk 1 above)
Comment (robertplatt-mo) [reply improver/cube_combiner.py:180]: @Anzerkhan27 Did you have a position on this?  I'm primarily wondering if the masked_add behaviour is in use anywhere and/or whether it could/should be deprecated in favour of the new approach which handles masked values on any of the operations.

File: improver/cube_combiner.py (hunk 2 above)
Comment (robertplatt-mo) [reply improver/cube_combiner.py]: Good spot. Done.

File: improver/cube_combiner.py [hunk 3]
```diff
@@ -228,6 +244,11 @@ def __init__(
             midpoint_bound:
                 If True, set the coordinate point to the midpoint of the bounds;
                 otherwise, use the upper bound. This is only used if expand_bound is also True.
+            preserve_mask_values:
```
Comment (MoseleyS) [inline improver/cube_combiner.py]: The behaviour I would prefer is

- by default, mask values are preserved (the PR continues to do this)
- when an option is provided to change this behaviour, the developer is specifying how the masked value should be interpreted. Something like `replace_masked_values_with = 0.0`

This would allow other instances of masked data to be treated differently. The possible mask-fill scenarios are 0, 1, large negative value, large positive value, NaN.

Comment (robertplatt-mo) [reply improver/cube_combiner.py]: We can do it that way that's no problem. We don't need the preserve_mask_values flag at all then in that scenario we can just replace values if a value is given for replace_masked_values_with.

I'll probably need to add tests for large negative/positive and NaN but I don't see it taking all that long.
